### PR TITLE
DM-5R / RD-5R version

### PR DESCRIPTION
**Pull request by ew1adg**
DM-5R support for OpenGD77 firmware.

### CODE REVIEW

File: firmware/source/user_interface/menuLanguage.c [hunk 1]
```diff
@@ -71,6 +71,10 @@ static void handleEvent(uiEvent_t *ev)
 		nonVolatileSettings.languageIndex = gMenusCurrentItemIndex;
 		currentLanguage = &languages[gMenusCurrentItemIndex];
 		menuSystemLanguageHasChanged();
+#if defined(PLATFORM_DM5R)
+		// Save settings on exit
+		settingsSaveSettings(false);
+#endif
```
Comment (f1rmb) [inline firmware/source/user_interface/menuLanguage.c]: Why saving the settings here ? Does the DM5R only have a real power switch without any MCU control ?

Comment (ew1adg) [reply firmware/source/user_interface/menuLanguage.c]: Yes. i was tried to find this circuits also with scanning all unused GPIO pins on this MCU.

Comment (ew1adg) [reply firmware/source/user_interface/menuLanguage.c]: In common, DM-5R schematics looks simpler, then other radios (GD-77, DM1801).
There is only single TX PA for VHF and UHF band.
Screenshot_20200331-003702_

**Comment by rogerclarkmelbourne** [comment]
I still have my RD-5R, I will look for the power control IC

**Comment by rogerclarkmelbourne** [comment]
GD-77 has this

image

**Comment by ew1adg** [comment]
It will be great. if it exists. May be i have a broken radio?
I got it in already disasembled state with signs of soldering in past.

**Comment by rogerclarkmelbourne** [comment]
I'm not sure if it is the same as the GD-77 and DM-1801

I have both a GD-77 and a RD-5R dissassembled at the moment, and I can try to compare.

The other possible way to know if there is a shutdown signal, is to attach a logic analyser to the EEPROM, because the firmware should write to the EEPROM when the radio is turned off. You can possibly also confirm this just using an oscilloscope, but as the I2C bus is shared with the AT1846 using a logic analyzer would be better

**Comment by rogerclarkmelbourne** [comment]
I think the power control is on the back of the PCB, not the side in the photo

**Comment by f1rmb** [comment]
Looking for the firmware shift value, I'm unable to find any firmware 2.1.7 for the DM-5R or RD-5R. The latest available seems to be 2.1.6. Do you have any download link for that version ?

**Comment by rogerclarkmelbourne** [comment]
I only have 2.1.6 which seems to be the latest on Baofeng's site.

Kai has some older versions in his github repo

https://github.com/talentraspel/GD-77/tree/master/Baofeng_RD-5R

Also in Kai's repo is 

https://github.com/talentraspel/GD-77/blob/master/GD-77_firmware_decrypt_encrypt_workdir/start_decrypt.bat

Which has the command line to decrypt the file

firmware_decrypt BF-5R_V2.1.6_BB8A_078001 datafile 0x7d54 decrypt

Which presumably means the offset for that version is 0x7D54

**Comment by ecolive19** [comment]
I might have an older firmware for the RD-5R.
I will check for it in the morning.
Eduardo.

**Comment by f1rmb** [comment]
Hi Roger,

   I know, I even redo the whole process, and it's matching what we already have, but the header is different from the one in this PR (has reported to be 2.1.7).

**Comment by rogerclarkmelbourne** [comment]
@f1rmb 

OK. We need to store a copy of 2.1.7 somewhere if it exists. Perhaps its DM-5R rather then RD-5R firmware.

Or we change to use the header from 2.1.6, as we have that version.

**Comment by f1rmb** [comment]
I search both, and all what I've found is 2.1.6.

**Comment by rogerclarkmelbourne** [comment]
Guys
I tried to understand the power circuit for this radio, and it does seem to be different from the GD-77

The switch in the volume control, seems to switch the positive power line from the battery, and the track on the PCB is quite wide, probably 3mm, which suggests this is the primary current supply to a lot of the radio.

However there is some circuit connected to the battery input , via some thin tracks, before the main power switch.

image
image
image

**Comment by rogerclarkmelbourne** [comment]
I have a few observations with the code.

It may be better to just defined some constants used for the graphics heights and use those in the code.

e.g. some places which had sizes set to 32 are set to 24 for this radio.

i.e a lot of places just need the height reduced by 8 pixels

I don't know if the clear rows code needs to be changed to FillRect etc, and perhaps just 1 row can be cleared rather than 2 ?

I know some places in the code the display needs to be changed in more complex ways, but I think at least 50% of the graphics changes could be made simpler by using constants.

But we'd need to name these so that the code is still readable.

e.g. perhaps GD77_POSITION32_RD5R_POSITION24

Or some other name ??

**Comment by rogerclarkmelbourne** [comment]
Saving the settings every time a button is pressed is going to place a load on the EEPROM, but it does have life of > 100,000 writes I think (possibly > 1M writes), so I guess it would be OK.

**Comment by f1rmb** [comment]
Hi Roger,

   I was thinking about identical thing, at the moment screen width/height are hardcoded, and the 8x8 rows could be a little more clean that way.

On the last picture, I guess it's the DC/DC converter (next to the coil) for the 3.3V.

**Comment by f1rmb** [comment]
There are some vias on the power track. Even, there is a thin track on top of the big R390 (but I can't see if it's connected to the power track).

**Comment by rogerclarkmelbourne** [comment]
Yes. The coil is probably part of the voltage regulator circuit.

But what concerns me is that the thick track on the PCB from the battery, (from the left of the image), goes only through the switch, apart from some very small tracks to the unknown IC on the left.

I don't see a good path for power apart from via the switch.

Perhaps  @ew1adg can comment on how he checked for any other paths for the power.

**Comment by rogerclarkmelbourne** [comment]
BTW. Perhaps I should make a new branch for DM5R-RD5R in my repo and then pull this version into that branch, so we could all work on it and raise PR's on the version in my repo.

I don't want to merge it into master yet.

**Comment by f1rmb** [comment]
Maybe there is a double mosfet on the bottom (there is a group a 4 vias), that shorting out the mecanical switch.

**Comment by rogerclarkmelbourne** [comment]
Yes. Just under the ribbon cable for the display are 2 solder dots, perhaps filled vias for extra current.

I will look on the other side of the PCB

**Comment by f1rmb** [comment]
Yes, creating a branch is better. At least we can clean/fix the whole UI stuff there.

**Comment by rogerclarkmelbourne** [comment]
Umm.  I will need to fix the merge conflicts, and then I will push the branch as the same name as this PR

**Comment by rogerclarkmelbourne** [comment]
I checked the other side of the PCB, but its in the middle of the RF section, so definitely no power control IC's

The PCB is 4 layer so it could be routed anywhere.

I'll need to use my resistance meter and perhaps I can find a connection somewhere on the PCB, but the location of the via is near the end of the PCB, so if it was connected anywhere, it would also be near the end of the PCB.

**Comment by f1rmb** [comment]
Okay Roger.

I'm still doubfull about no power control from the MCU, as if the user switch the power OFF at anytime, e.g. in a middle of an EEPROM write, that can corrupt the data, and create a nice surprise on the next boot.

**Comment by ew1adg** [comment]
> I have a few observations with the code.
> 
> It may be better to just defined some constants used for the graphics heights and use those in the code.
> 
> e.g. some places which had sizes set to 32 are set to 24 for this radio.
> 
> i.e a lot of places just need the height reduced by 8 pixels
> 
> I don't know if the clear rows code needs to be changed to FillRect etc, and perhaps just 1 row can be cleared rather than 2 ?
> 
> I know some places in the code the display needs to be changed in more complex ways, but I think at least 50% of the graphics changes could be made simpler by using constants.
> 
> But we'd need to name these so that the code is still readable.
> 
> e.g. perhaps GD77_POSITION32_RD5R_POSITION24
> 
> Or some other name ??

Thank you. May be better to move screen layout related things to separate header file, to avoid tons of preprocessor section?
example:
include/user_interface/layout/128x64.h
include/user_interface/layout/128x48.h

and check #ifdef once, maybe twice.

**Comment by ew1adg** [comment]
i am offer to save setting to EEPROM at exit of options menu only.
for saving VFO parameter use already existing feature (SK1+SK2)

**Comment by ew1adg** [comment]
> Okay Roger.
> 
> I'm still doubfull about no power control from the MCU, as if the user switch the power OFF at anytime, e.g. in a middle of an EEPROM write, that can corrupt the data, and create a nice surprise on the next boot.

Today i was take another bruteforce attemp to find powercontrol GPIO.
There is no power control GPIO was found.

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

How do you brute force the power on?

Set all GPIO to true then turn off the power ?

**Comment by ew1adg** [comment]
> @ew1adg
> 
> How do you brute force the power on?
> 
> Set all GPIO to true then turn off the power ?

Yes, only one difference, i am was test one GPIO pin at one run.

**Comment by ew1adg** [comment]
port A
port B
port C
port D
port E

Known GPIOs. Ports A-E.

**Comment by rogerclarkmelbourne** [comment]
If the MCU is controlling the power there also needs to be an input that senses the that the power has been turned off.

So another way to brute force this, is to set all pins to input, then power the MCU from the debugger port via a diode, then turn off the battery power and observe whether any inputs have changed.

**Comment by rogerclarkmelbourne** [comment]
I've pulled and merged this PR into a new branch

https://github.com/rogerclarkmelbourne/OpenGD77/tree/ew1adg-dm5r_support

Note. I had to manually port the whole of  uiVFOMode as manually merging was more difficult than changing the calls to ucPrint etc.


So that there does not need to be so many conditional compilation statements for the position of the graphics, perhaps we can have some constants for the ROW.
Because often the Y position is a multiple of 16 (or sometimes a multiple of 8)

I know sometimes we position things up or down 1 or 2 pixels from the start of the row, but if we define some constants for this , we can still add or subtract values, and hopefully the compiler should optimise out these calculations.

The other thing is the font size. The DM5R code seems to mainly use the 8x8 font instead of the 8x16 font.

I see the fonts sizes are an enum

```
typedef enum
{
	FONT_6x8 = 0,
	FONT_6x8_BOLD,
	FONT_8x8,
	FONT_8x16,
	FONT_16x32
} ucFont_t;
```

and that printCore gets the pointer to the font data via a switch statement.

So if we rename the fonts to perhaps just

```
typedef enum
{
	FONT_EXTRA_SMALL= 0,
	FONT_EXTRA_SMALL_BOLD,
	FONT_SMALL,
	FONT_MEDIUM,
	FONT_LARGE
} ucFont_t;
```

Then we can change printCore via an ifdef so that the actual font data for each enum is different

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

BTW.

I am leaving this open, while I still work on the changes I want to make, so that there are less differences for the DM-5R

I now have a branch for the DM-5R version, which I am slowly updating

https://github.com/rogerclarkmelbourne/OpenGD77/tree/ew1adg-dm5r_support

When I am happy what the changes are minimized I will merge it into master.

**Comment by ew1adg** [comment]
Hello, at this time i am doing similar job. Make a font mapping for lower fonts on DM-5R.
And reviewing all changes done before.

**Comment by rogerclarkmelbourne** [comment]
OK

I think I will be working on this for 1 or 2 more days and then I will release it for beta testing to the forum https://www.opengd77.com
I also know one other ham who has this radio

**Comment by ew1adg** [comment]
Font mapping looks like this:
at this case i will unable to use 16x32 font, that used to batttery voltage indication.

pr2

pr1

**Comment by rogerclarkmelbourne** [comment]
My font mapping is similar, except the names are just 

FONT_SIZE_1
FONT_SIZE_1_BOLD
FONT_SIZE_2 
etc

I think this is more flexible / generic

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

Thanks

I already made most of these changes into my own branch of your code, but the font names I used were different because they are more generic

I will review your other changes, and manually edit the same changes into my branch as appropriate

**Comment by ew1adg** [comment]
**rogerclarkmelbourne**, please look at this changes, that i was recently made.
Highlights:
- huge reduce of usage preprocessor`s directives
- revert to 3-line main menu

**Comment by ew1adg** [comment]
> @ew1adg
> 
> Thanks
> 
> I already made most of these changes into my own branch of your code, but the font names I used were different because they are more generic
> 
> I will review your other changes, and manually edit the same changes into my branch as appropriate

Ok. thank you! 73

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

Did you discover whether the official firmware has some other way to save the nonVolatile settings apart from saving immediately ?

**Comment by ew1adg** [comment]
@rogerclarkmelbourne At his moment it is not possible at all. We went to the village with my family for self isolation. There is no any equipment to discover internal word of this radio.

**Comment by rogerclarkmelbourne** [comment]
BTW. I have moved the torch LED function into LEDs.c  I don't think its necessary for an extra file for one LED

**Comment by rogerclarkmelbourne** [comment]
Ok. No worries.

Its better to be in isolation.

I have not gone outside our front gate of the garden for 7 days.

**Comment by ew1adg** [comment]
@rogerclarkmelbourne 
Today i was noticed a stange behavior with DMR Rx audio.
When i set Rx and Tx freq. to 144.5375 (My hotspot freq.) the rx audio will work fine, but when i was move Tx freq audio upper, for example to 146.0000, Rx audio was muted at all.

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

I can only receive from my hotspot.  I think the calibration may be incorrect in my radio. 
I need to do some more tests when I have time

**Comment by rogerclarkmelbourne** [comment]
DM-5R/ RD-5R support has now been merged into the Master branch

It probably needs more work but it is mostly working.

I will leave this PR open,  in case I need to check any part of the PR

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

Does this radio have a separate RF PA for VHF and UHF?

I thought it only had 1 RF PA section, but you use PE6 for the UHF PA and PC4 for the VHF PA

Also did you confirm that the VHF and UHF Rx amplifier control pins are swapped?

I'm having problems waking a repeater. If I press and hold the PTT it does not wake. But if I press and immediately release the PTT the repeater wakes.

This can be caused if the Rx amp is not being enabled during the WakeUp sequence

Also, when you transmit, do the LED behind the keypad flash ever 30mS because of the DMR pulses ? This happens on my radio :-(

Thankyou

**Comment by ew1adg** [comment]
There is only one Tx RF amp. in this radio.

**Comment by ew1adg** [comment]
Backlit diodes flashing?
It is visible only keyboard, but very small drop in brightness.

**Comment by rogerclarkmelbourne** [comment]
OK. I think because I have the front removed from the radio, the flickering is more obvious

BTW. 
I asked VK7ZJA, who has analysed this radio, and he told me that although there is only one PA there are switchable RF output filters.

**Comment by ew1adg** [comment]
It will be great to get a help. 
Do you have a backup of 512kbit I2C eeprom?
I have issue with a contact list, but i have only backup of SPI flash.

**Comment by rogerclarkmelbourne** [comment]
[EEPROM.zip](https://github.com/rogerclarkmelbourne/OpenGD77/files/4485198/EEPROM.zip)
This is my backup. It probably has some part of a very basic codeplug

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

Does this radio have a separate baseband RF chip to allow it to receive broadcast radio, 68Mhz - 108Mhz

Is that what the "Band" button controls ?

**Comment by rogerclarkmelbourne** [comment]
@ew1adg 

Does the RD-5R have the RDA5802  chip in it.

For 68 - 108Mhz??

**Comment by rogerclarkmelbourne** [comment]
I will close this now as it has been manually merged

Use this issue for all further comments

https://github.com/rogerclarkmelbourne/OpenGD77/issues/472

**Comment by rogerclarkmelbourne** [comment]
Dzmitry

FYI

The Tx power in your original version is very wrong.

There seems to be a big difference between the PA control on the RD-5R / DM-5R and the GD-77 or DM-1801.

On the GD-77 and DM-1801 the value from the calibration table is multipled by 16, and this value is used for the PA drive DAC.

On the DM-5R, sending 16 x the calibration value results in very high output power even on the 1W  power setting, which uses the "Low" power value from the calibration table.

I tried different multiplier values on the RD-5R, but I could not get the Tx power to match the value produced with the official version on Low or high power..

e.g. On 145Mhz, the official firmware outputs 1.5W, and High power, outputs 4.2W

But I tried multiplers of 4,5,6 etc and none of these multipliers resulted in the same output power.

So I also tried changing the AT1846S output power setting. This is read from the calibration table and is normally 8.

But changing this value and also the multiplier for the PA drive DAC never results in 1.5W on Low and 4.2W on High power in this firmware.

Also, more worrying... On UHF with this firmware the power decreases by around 5% per second, which suggests perhaps that the PA FET is heating a lot, caused by bad SWR, possibly caused by some incorrect RF setting.

Are you sure you configured the Tx PA filter correctly for UHF and VHF or is there a chance that these are transposed ?

I may try transposing the UHF and VHF PA control pins and see if this makes any difference to the operation.

**Comment by ew1adg** [comment]
Hi, permamently high power also on a VHF too?

**Comment by rogerclarkmelbourne** [comment]
Yes. VHF and UHF

I tested my radio and Tx power on VHF and UHF was 8W, on the 1W setting which uses the "Low" value from the calibration table

Also, on UHF the power output is not stable.

**Comment by ew1adg** [comment]
May be an analog part similar to UV-5R? 
http://www.radioscanner.ru/files/download/file14137/shema-baofeng-5r.pdf

**Comment by rogerclarkmelbourne** [comment]
OK.

Do you have a power meter ?

**Comment by ew1adg** [comment]
> May be an analog part similar to UV-5R?
> http://www.radioscanner.ru/files/download/file14137/shema-baofeng-5r.pdf
> For next 2-3 days i will investigate this by drawing schematics from radio PCB board.

No, at this location.

**Comment by rogerclarkmelbourne** [comment]
OK

No worries.

**Comment by rogerclarkmelbourne** [comment]
FYI see

https://github.com/rogerclarkmelbourne/OpenGD77/issues/602

I think we are using the wrong calibration data address. This will also have a problem on the power levels !

**Comment by rogerclarkmelbourne** [comment]
See the latest push to the development branch.

The problem was the calibration table is not at 0x17c00 its at 0x8f000

See 

https://www.opengd77.com/viewtopic.php?f=13&t=827&p=8310#p8309

**Comment by rogerclarkmelbourne** [comment]
I will close this again, as there are no current bugs with the RD-5R / DM-5R version

**Pull request closed without merging**